### PR TITLE
Dataloader multilevel dispatch

### PR DESCRIPTION
**Pull request by andimarek**
this is an attempt to improve the performance of DataLoader for lists in lists: basically changing the ExecutionStrategy to a breadth deep traversal to batch load the maximum possible

### CODE REVIEW

File: src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentation.java [hunk 1]
```diff
@@ -84,34 +82,32 @@ private void setAggressivelyBatching(boolean aggressivelyBatching) {
...
+            outstandingFieldFetchCounts.put(level, count);
         }
 
-        private void exitList() {
-            synchronized (this) {
-                if (!stack.isEmpty()) {
-                    stack.pop();
-                }
-            }
+        // TODO: should be threadsafe
```
Comment (bbakerman) [inline src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentation.java]: To make this threadsafe - invent a new type that encapsulates level + count and make that have read / write locks.

File: src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentation.java [hunk 2]
```diff
@@ -122,11 +118,38 @@ public InstrumentationState createState() {
...
+
+        List<String> dlKeysWithData = dataLoaderRegistry.getKeys().stream().filter(key -> dataLoaderRegistry.getDataLoader(key).dispatchDepth() > 0).collect(Collectors.toList());
+        if (!dlKeysWithData.isEmpty()) {
+            System.out.println("\t\tDispatched!");
+        }
+        dlKeysWithData.forEach(key -> {
+            System.out.println(String.format("'%s' has %d depth", key, dataLoaderRegistry.getDataLoader(key).dispatchDepth()));
+        });
+
+        System.out.println("\n\n");
```
Comment (bbakerman) [inline src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentation.java]: Clearly the printfs need to go at some stage!

**Comment by bbakerman** [comment]
@andimarek  - I split the logic contained in the instrumentation class out into its own `graphql.execution.instrumentation.dataloader.FieldLevelTrackingApproach` class

This will allow us to include Sam Bs Braid code as another approach.  We can then continue to test with both say and work out which is the better one.  We might then retain one or the other or maybe both.

**Comment by andimarek** [comment]
breaking change because of the change to the return type of `ExecutionStrategyInstrumentationContext.beginExecutionStrategy`.

Also because of the changed return values to `completeValue` etc.: they return now a FieldValueInfo.

**Pull request merged**